domain/secret/service/watcher.go

```diff
@@ -155,6 +150,237 @@ func (s *WatchableService) WatchSecretRevisionsExpiryChanges(_ context.Context,
+		changestream.All,
+		query,
+	)
+	if err != nil {
+		return internalerrors.Capture(err)
+	}
+
+	table, query = w.state.InitialWatchStatementForObsoleteRevision(w.appOwners, w.unitOwners)
+	w.revisionWatcher, err = w.watcherFactory.NewNamespaceWatcher(
+		table,
+		changestream.Changed,
+		query,
+	)
+	return internalerrors.Capture(err)
+}
+
+func (w *obsoleteWatcher) scopedContext() (context.Context, context.CancelFunc) {
+	return context.WithCancel(w.catacomb.Context(context.Background()))
+}
+
+func splitSecretRevision(s string) (string, int) {
+	parts := strings.Split(s, "/")
+	if len(parts) < 2 {
+		return parts[0], 0
+	}
+	rev, _ := strconv.Atoi(parts[1])
+	return parts[0], rev
+}
+
+func (w *obsoleteWatcher) mergeSecretChanges(
+	ctx context.Context, currentChanges []string, receivcedSecretIDs []string,
+) ([]string, error) {
+	if len(receivcedSecretIDs) == 0 {
+		return currentChanges, nil
+	}
+
+	// pushChanges pushes the secret ID to the changes slice.
+	// At the same time, any previously added obsolete revisions of this secret are removed from the slice.
+	pushChanges := func(secretID string) {
+		currentChanges = slices.DeleteFunc(currentChanges, func(s string) bool {
+			id, _ := splitSecretRevision(s)
+			return id == secretID
+		})
+		currentChanges = append(currentChanges, secretID)
+	}
+
+	w.logger.Criticalf(ctx, "secretWatcher(%v-%v): received changes %v", w.appOwners, w.unitOwners, receivcedSecretIDs)
+	for _, uriStr := range receivcedSecretIDs {
+		uri, err := coresecrets.ParseURI(uriStr)
+		if err != nil {
+			return currentChanges, errors.Trace(err)
+		}
+		owned, err := w.state.IsSecretOwnedBy(ctx, uri, w.appOwners, w.unitOwners)
+		w.logger.Criticalf(ctx, "secretWatcher(%v-%v): uri %q, owned %v", w.appOwners, w.unitOwners, uri, owned)
+		if err != nil && !errors.Is(err, secreterrors.SecretNotFound) {
+			return currentChanges, errors.Trace(err)
+		}
+		if owned {
+			// It's still owned, so the event must be triggered by creation.
+			// Ensure we are tracking the secret URI.
+			w.knownSecretURIs.Add(uri.ID)
+
+			// We are only interested in a previously owned secret that has been deleted,
+			// so ignore this one and continue.
+			continue
+		}
+
+		if !owned && w.knownSecretURIs.Contains(uri.ID) {
+			// An onwed secret has been deleted, we need to notify the URI change.
+			pushChanges(uri.ID)
+
+			// No need to track this one anymore.
+			w.knownSecretURIs.Remove(uri.ID)
+		}
+	}
+	return currentChanges, nil
+}
+
+func (w *obsoleteWatcher) mergeRevisionChanges(
+	ctx context.Context, currentChanges []string, newChanges []string,
+) ([]string, error) {
+	if len(newChanges) == 0 {
+		return currentChanges, nil
+	}
+
+	w.logger.Criticalf(ctx, "revisionWatcher(%v-%v): received changes %v", w.appOwners, w.unitOwners, newChanges)
+	// We are receiving all the obsolete revision UUIDs changes in the model, so we need to filter
+	// only the ones that are owned.
+	revisionIDs, err := w.state.GetRevisionIDsForObsolete(
```

Comment:
To confirm - this query will exclude any revisions whose parent secret has been deleted right? Do we have unit test coverage for that.

Reply:
When a secret is deleted, all its relevant resources(including obsolete revisions) are deleted.

Reply:
We could be processing obsolete revision changes and then the secret is deleted. So we need to account for that potential race. So just wanting to confirm the Get method does the right thing and is it tested.